### PR TITLE
PIM-4407: add comparators

### PR DESCRIPTION
**Pull request by skeleton**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | ?
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-4407
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               | N

### CODE REVIEW

File: spec/Pim/Component/Catalog/Comparator/BooleanComparatorSpec.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        $this->getChanges($changes, $originals)->shouldReturn($changes);
+
+        $changes = ['value' => 1, 'locale' => 'en_US', 'scope' => 'ecommerce'];
+        $originals = [];
+
+        $this->getChanges($changes, $originals)->shouldReturn($changes);
+    }
+
+    function it_get_changes_when_changing_value()
```
Comment (solivier) [inline spec/Pim/Component/Catalog/Comparator/BooleanComparatorSpec.php]: gets

File: spec/Pim/Component/Catalog/Comparator/BooleanComparatorSpec.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $this->shouldBeAnInstanceOf('Pim\Component\Catalog\Comparator\ComparatorInterface');
+    }
+
+    function it_supports_comparison()
+    {
+        $this->supportsComparison('pim_catalog_boolean')->shouldBe(true);
+    }
+
+    function it_get_changes_when_adding_value()
```
Comment (solivier) [inline spec/Pim/Component/Catalog/Comparator/BooleanComparatorSpec.php]: gets

File: spec/Pim/Component/Catalog/Comparator/MediaComparatorSpec.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    function it_supports_media_type()
+    {
+        $this->supportsComparison('pim_catalog_file')->shouldReturn(true);
+        $this->supportsComparison('pim_catalog_image')->shouldReturn(true);
+        $this->supportsComparison('other')->shouldBe(false);
+    }
+
+    function it_get_changes_when_add_file()
```
Comment (solivier) [inline spec/Pim/Component/Catalog/Comparator/MediaComparatorSpec.php]: gets

File: src/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterComparatorsPass.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterComparatorsPass.php]: wrong

File: src/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterComparatorsPass.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+namespace Pim\Bundle\CatalogBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Compiler pass that register product draft value presenters into the product draft twig extension
+ *
+ * @author Gildas Quemener <gildas@akeneo.com>
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterComparatorsPass.php]: I think copyrights and correct license is missing

File: src/Pim/Component/Catalog/Comparator/MediaComparator.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * @param string $filePath
+     *
+     * @return null|string
+     */
+    protected function getHashFile($filePath = null)
+    {
+        if (null !== $filePath) {
```
Comment (solivier) [inline src/Pim/Component/Catalog/Comparator/MediaComparator.php]: return null !== $filePath ? sha1_file($filePath) : null;

File: src/Pim/Component/Catalog/Comparator/ComparatorInterface.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Get the changes between a normalized product value instance and the updated data
+     * If no changes detected, then the method returns null
+     *
+     * @param array $data
+     * @param array $originals
+     *
+     * @return array|null
+     */
+    public function getChanges(array $data, array $originals);
```
Comment (nidup) [inline src/Pim/Component/Catalog/Comparator/ComparatorInterface.php]: Still weird for me to have ComparatorInterface with a main method getChanges() 
A compare() method to return the diff could be ok no?

**Pull request merged**